File: nodecore-spv/src/main/java/veriblock/wallet/PendingTransactionDownloadedListenerImpl.java
package veriblock.wallet;

import org.veriblock.sdk.models.Sha256Hash;
import veriblock.SpvContext;
import veriblock.listeners.PendingTransactionDownloadedListener;
import veriblock.model.StandardTransaction;
import veriblock.model.TransactionMeta;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.locks.ReentrantLock;

public class PendingTransactionDownloadedListenerImpl implements PendingTransactionDownloadedListener {

    private final SpvContext spvContext;

    private Map<Sha256Hash, StandardTransaction> transactions;
    private final Ledger ledger = new Ledger();

    private final KeyRing keyRing = new KeyRing();
    private final ReentrantLock lock = new ReentrantLock(true);

    public PendingTransactionDownloadedListenerImpl(SpvContext spvContext) {
        this.spvContext = spvContext;
    }

    @Override
    public void onPendingTransactionDownloaded(StandardTransaction transaction) {

    }

    void loadTransactions(List<StandardTransaction> toLoad) {
        for (StandardTransaction tx : toLoad) {
            transactions.put(tx.getTxId(), tx);
            spvContext.getTransactionPool().insert(tx.getTransactionMeta());
        }
    }

    public void commitTx(StandardTransaction tx) {
        lock.lock();
        try {
            if (transactions.containsKey(tx.getTxId())) {
                return;
            }

            tx.getTransactionMeta().setState(TransactionMeta.MetaState.PENDING);
            transactions.put(tx.getTxId(), tx);

            ledger.record(tx);
        } finally {
            lock.unlock();
        }
    }

    public StandardTransaction getStandardTransaction(Sha256Hash txId) {
        return transactions.get(txId);
    }

    public Collection<StandardTransaction> getStandardTransactions() {
        return Collections.unmodifiableCollection(transactions.values());
    }

    private boolean isTransactionRelevant(StandardTransaction tx) {
        if (this.transactions.containsKey(tx.getTxId())) {
            return true;
        }

        if (keyRing.contains(tx.getInputAddress().get())) {
            return true;
        }

        return tx.getOutputs().stream().anyMatch(output -> keyRing.contains(output.getAddress().get()));
    }

    private void addTransaction(StandardTransaction tx) {
        lock.lock();
        try {
            transactions.putIfAbsent(tx.getTxId(), tx);
            ledger.record(tx);
//            save();
        } finally {
            lock.unlock();
        }
    }
}

File: nodecore-spv/src/main/java/veriblock/wallet/PendingTransactionDownloadedListenerImpl.kt
package veriblock.wallet

import org.veriblock.core.crypto.Sha256Hash
import veriblock.SpvContext
import veriblock.listeners.PendingTransactionDownloadedListener
import veriblock.model.Output
import veriblock.model.StandardTransaction
import veriblock.model.TransactionMeta
import java.util.Collections
import java.util.concurrent.locks.ReentrantLock

class PendingTransactionDownloadedListenerImpl(
    private val spvContext: SpvContext
) : PendingTransactionDownloadedListener {
    private val transactions: MutableMap<Sha256Hash?, StandardTransaction>? = null
    private val ledger = Ledger()
    private val keyRing = KeyRing()
    private val lock = ReentrantLock(true)
    override fun onPendingTransactionDownloaded(transaction: StandardTransaction?) {}

    fun loadTransactions(toLoad: List<StandardTransaction>) {
        for (tx in toLoad) {
            transactions!![tx.txId] = tx
            spvContext.transactionPool.insert(tx.transactionMeta!!)
        }
    }

    fun commitTx(tx: StandardTransaction) {
        lock.lock()
        try {
            if (transactions!!.containsKey(tx.txId)) {
                return
            }
            tx.transactionMeta!!.setState(TransactionMeta.MetaState.PENDING)
            transactions[tx.txId] = tx
            ledger.record(tx)
        } finally {
            lock.unlock()
        }
    }

    fun getStandardTransaction(txId: Sha256Hash?): StandardTransaction? {
        return transactions!![txId]
    }

    fun getStandardTransactions(): Collection<StandardTransaction> =
        Collections.unmodifiableCollection(transactions!!.values)

    private fun isTransactionRelevant(tx: StandardTransaction): Boolean {
        if (transactions!!.containsKey(tx.txId)) {
            return true
        }
        return if (keyRing.contains(tx.inputAddress!!.get())) {
            true
        } else {
            tx.getOutputs().any {
                keyRing.contains(it.address.get())
            }
        }
    }

    private fun addTransaction(tx: StandardTransaction) {
        lock.lock()
        try {
            transactions!!.putIfAbsent(tx.txId, tx)
            ledger.record(tx)
            //save();
        } finally {
            lock.unlock()
        }
    }

}
